### PR TITLE
CredentialServer: extract code dealing with credential helpers

### PR DESCRIPTION
**Pull request by mook-as**
This extracts out the code to run credential helpers so that it can be reused elsewhere.

This will be used for #4362.

### CODE REVIEW

**Comment by ericpromislow** [comment]
Initial unit test failure (on macOS):

```
Error: expect(received).resolves.toEqual()

Received promise rejected instead of resolved
Rejected to value: [TypeError: Cannot read properties of undefined (reading 'isPackaged')]

    at expect (/Users/ericp/workspace/rancher/desktop/node_modules/expect/build/index.js:178:15)
    at Object.<anonymous> (/Users/ericp/workspace/rancher/desktop/pkg/rancher-desktop/main/credentialServer/__tests__/credentialUtils.spec.ts:44:11)
    at Promise.then.completed (/Users/ericp/workspace/rancher/desktop/node_modules/jest-circus/build/utils.js:391:28)
 ```

**Review by ericpromislow — CHANGES_REQUESTED** [review]
Unit tests are failing the same way on both macos and windows. Everything else looks good

**Comment by mook-as** [comment]
Sorry about that, I missed a dependency when cherry-picking from my work branch. It should be fine now.

File: pkg/rancher-desktop/main/credentialServer/__tests__/credentialUtils.spec.ts [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+  });
+
+  it('runs the command', async() => {
+    const expected = `Some output`;
+
+    jest.spyOn(fs.promises, 'readFile').mockImplementation((filepath, options) => {
+      const home = findHomeDir() ?? '';
+
+      expect(filepath).toEqual(path.join(home, '.docker', 'config.json'));
+      expect(options).toMatchObject({ encoding: 'utf-8' });
```
Comment (ericpromislow) [inline pkg/rancher-desktop/main/credentialServer/__tests__/credentialUtils.spec.ts:28]: Does this line matter?

File: pkg/rancher-desktop/main/credentialServer/__tests__/credentialUtils.spec.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+      return Promise.reject(error);
+    });
+
+    jest.mocked(spawnFile).mockImplementation(() => Promise.resolve({}));
+
+    await expect(runCommand('pika')).rejects.toBe(error);
+    expect(jest.mocked(spawnFile)).not.toHaveBeenCalled();
+  });
+
+  describe.each([
```
Comment (ericpromislow) [inline pkg/rancher-desktop/main/credentialServer/__tests__/credentialUtils.spec.ts:64]: The nested loops obscure what's actually being tested. Could you put some sample tests in the comments above the `describe.each` block so it's clearer.

I wouldn't object to making this a general practice, but in this case the main line of interest is about 25 lines below the start of the first loop.

File: pkg/rancher-desktop/main/credentialServer/__tests__/credentialUtils.spec.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+describe('list', () => {
+  let config: { credsStore: string, credHelpers?: Record<string, string>} = { credsStore: 'unset' };
+  let helpers: Record<string, any> = {};
+
+  beforeEach(() => {
+    jest.spyOn(fs.promises, 'readFile').mockImplementation((filepath, options) => {
+      const home = findHomeDir() ?? '';
+
+      expect(filepath).toEqual(path.join(home, '.docker', 'config.json'));
+      expect(options).toMatchObject({ encoding: 'utf-8' });
```
Comment (ericpromislow) [inline pkg/rancher-desktop/main/credentialServer/__tests__/credentialUtils.spec.ts:117]: Again I'm not sure what checking the options field adds to this test. Maybe if we were handling `ucs-2` on windows...

**Review by ericpromislow — CHANGES_REQUESTED** [review]
The code looks fine, except I don't know why we need to check `options`. Just requesting a couple more comments.

File: pkg/rancher-desktop/main/credentialServer/__tests__/credentialUtils.spec.ts (hunk 1 above)
Comment (mook-as) [reply pkg/rancher-desktop/main/credentialServer/__tests__/credentialUtils.spec.ts]: Not really, no.  I'll remove it.

File: pkg/rancher-desktop/main/credentialServer/__tests__/credentialUtils.spec.ts (hunk 3 above)
Comment (mook-as) [reply pkg/rancher-desktop/main/credentialServer/__tests__/credentialUtils.spec.ts]: Also removed.

File: pkg/rancher-desktop/main/credentialServer/__tests__/credentialUtils.spec.ts (hunk 2 above)
Comment (mook-as) [reply pkg/rancher-desktop/main/credentialServer/__tests__/credentialUtils.spec.ts:66]: Comments added; hopefully that makes things clearer.

**Review by ericpromislow — APPROVED** [review]
LG

**Pull request merged**